ReadMe.md
# Paintings++

README.md by [Joel Murphy](#credits)

_Paintings++_, now with

 _Painting Selection GUI Revamped_ built in, 

## Just the GUI

If you only want to use the selection GUI, there is no need to configure anything. Just install the mod through Twitch, or by [manual installation][link_install_mod], and you're all set.

## Art Collections

_Paintings++_ uses a **texture**, which is a single PNG image, divided up into cells, with each **cell** representing one picture; and a **pattern**, describing how the image is divided. We call this pairing, of texture and pattern, an **art collection**.

The _**vanilla**_ art collection is selected by default on a new install of _Paintings++_. To change to a different collection, you will first need the configuration file, `morepaintings.cfg` to be created. Running Minecraft after you've installed the _Paintings++_ `.jar` file, the configuration file will be automatically created in the `config` subfolder of your Forge installation. If you need help finding the Forge folder, see the _manual installation_ link under [Just the GUI](#just-the-gui).

Open the configuration file, `morepaintings.cfg`, and enter the name of the collection under `painting mode`. Valid names for built-in patterns are _**gibea**_, _**sphax**_, _**tinypics**_, _**mediumpics**_, _**insane**_, _**new_insane**_, _**extended**_, and _**massive**_. You can also set the name to _**vanilla**_ or to the name of a [custom art collection](#custom-art-collections). After editing the config, if Minecraft is still running, you'll need to close Minecraft and restart.

If you don't want to add a texture file, the _**gibea**_ collection comes preloaded with a complete set of paintings. Otherwise, you'll need a `.png` file containing all of your pictures. You can find a few contributed 

### Built-in Patterns

Eight collections of varying size, and quantity of pictures, are built into _Paintings++_. Some of the collections come with default images in each cell and, wherever there isn't a default image, there's a solid color for each painting, based on its size and shape.

Below is a table summarizing the built-in collection. Clicking on the collection name will expand the row to show a thumbnail of the texture template. Click either the template, or the download symbol (<span class="fa fa-download"> </span>), to download a texture.

| Collection                                                                                            | Resolution<sup>1</sup> | # Cells | # Pictures | Pattern Size<sup>2</sup> |                                                          |
| :---------------------------------------------------------------------------------------------------- | :--------------------: | :-----: | :--------: | :----------------------: | :------------------------------------------------------: |
| <details><summary>gibea</summary>[![gibea.png][link_gibea]][link_gibea]</details>                     | 256&times;256          | 88      | 88         | 16&times;16              | [<span class="fa fa-download"> </span>][link_gibea]      |
| <details><summary>sphax</summary>[![sphax.png][link_sphax]][link_sphax]</details>                     | 256&times;256          | 50      | 26         | 16&times;16              | [<span class="fa fa-download"> </span>][link_sphax]      |
| <details><summary>tinypics</summary>[![tinypics.png][link_tinypics]][link_tinypics]</details>         | 512&times;512          | 302     | —          | 32&times;32              | [<span class="fa fa-download"> </span>][link_tinypics]   |
| <details><summary>mediumpics</summary>[![mediumpics.png][link_mediumpics]][link_mediumpics]</details> | 512&times;512          | 222     | —          | 32&times;32              | [<span class="fa fa-download"> </span>][link_mediumpics] |
| <details><summary>new_insane</summary>[![new_insane.png][link_new_insane]][link_new_insane]</details> | 512&times;512          | 121     | 42         | 32&times;32              | [<span class="fa fa-download"> </span>][link_new_insane] |
| <details><summary>insane</summary>[![insane.png][link_insane]][link_insane]</details>                 | 512&times;512          | 121     | 26         | 32&times;32              | [<span class="fa fa-download"> </span>][link_insane]     |
| <details><summary>extended</summary>[![extended.png][link_extended]][link_extended]</details>         | 512&times;512          | 171     | 26         | 32&times;32              | [<span class="fa fa-download"> </span>][link_extended]   |
| <details><summary>massive</summary>[![massive.png][link_massive]][link_massive]</details>             | 1008&times;1008        | 281     | —          | 63&times;63              | [<span class="fa fa-download"> </span>][link_massive]    |

<small>

1. These reference textures use 16&times;16 pixels per block but higher detail paintings can be made by creating a larger texture. Integer multiples of 16&times;16 are recommended (e.g. 32&times;32, 64&times;64, and so on) for each block.
2. Pattern size is in blocks.

</small>

_**Q: Do you mean I have to buy something like Adobe Photoshop?**_  
_**A:**_ Not at all. There are numerous free image editors such as [Krita](https://krita.org/), [GIMP](https://www.gimp.org/), and [Paint.net](https://www.getpaint.net/). If you'd like to draw some of your own pixel art, [Pixilart](https://www.pixilart.com/) has a pretty neat editor that works right inside your web browser.

### Custom Patterns

## Useless Tables

### Sizes in Default Templates

This table shows the distribution of painting sizes in each collection aswell as the symbol used for each size. There is no requirement that you use the symbols we did, only that each symbol have only one size.

| Symbol | Size  | Area | Gibea | Sphax | Tinypics | Mediumpics | Insane<sup>1</sup> | Extended | Massive |
| :----: | :---: | :--: | :---: | :---: | :------: | :--------: | :----------------: | :------: | :-----: |
| A      | 1x1   | 1    | 32    | 18    | 76       | 48         | 11                 | 32       | 20      |
| B      | 1x2   | 2    | 12    | 4     | 32       | 10         | 11                 | 12       | 16      |
| C      | 1x3   | 3    |       | 3     | 32       | 5          | 6                  | 8        | 5       |
| D      | 1x4   | 4    |       |       | 4        |            |                    |          | 3       |
| E      | 1x10  | 10   |       |       |          |            |                    |          | 1       |
| F      | 2x1   | 2    | 38    | 12    | 32       | 24         | 14                 | 28       | 20      |
| G      | 2x2   | 4    | 11    | 12    | 36       | 64         | 18                 | 12       | 1       |
| H      | 2x3   | 6    |       | 1     | 16       | 16         | 6                  | 16       | 11      |
| I      | 2x4   | 8    |       |       | 4        | 4          |                    | 5        | 8       |
| J      | 3x1   | 3    |       | 2     | 32       |            | 6                  | 8        | 5       |
| K      | 3x2   | 6    |       | 1     | 16       | 18         | 2                  | 16       | 19      |
| L      | 3x3   | 9    |       |       | 8        | 9          | 3                  | 8        | 7       |
| M      | 3x4   | 12   |       |       |          |            |                    | 6        | 7       |
| N      | 3x5   | 15   |       |       |          |            |                    |          | 5       |
| O      | 3x6   | 18   |       |       |          |            |                    |          | 4       |
| P      | 3x7   | 21   |       |       |          |            |                    |          | 2       |
| Q      | 4x1   | 4    |       | 1     | 8        |            | 8                  | 4        | 12      |
| R      | 4x2   | 8    | 1     | 1     | 4        | 10         | 5                  | 5        | 18      |
| S      | 4x3   | 12   | 2     | 4     | 5        | 16         | 8                  | 12       | 20      |
| T      | 4x4   | 16   | 3     | 3     | 3        | 8          | 10                 | 3        | 6       |
| U      | 4x5   | 20   |       |       |          |            |                    |          | 1       |
| V      | 4x6   | 24   |       |       |          |            | 4                  | 4        | 16      |
| W      | 4x8   | 32   |       |       |          |            | 2                  |          | 4       |
| X      | 5x1   | 5    |       |       |          |            |                    |          | 5       |
| Y      | 5x5   | 25   |       | 1     |          |            | 2                  |          | 7       |
| Z      | 5x7   | 35   |       |       |          |            |                    |          | 5       |
| a      | 5x8   | 40   |       |       |          |            |                    |          | 5       |
| b      | 5x10  | 50   |       |       |          |            |                    |          | 2       |
| c      | 6x3   | 18   |       |       |          |            |                    |          | 6       |
| d      | 6x4   | 24   |       |       |          |            |                    | 4        | 10      |
| e      | 6x9   | 54   |       |       |          |            |                    |          | 3       |
| f      | 7x5   | 35   |       |       |          |            |                    |          | 4       |
| g      | 8x1   | 8    |       |       |          |            | 1                  |          |         |
| h      | 8x2   | 16   |       |       |          |            |                    |          | 5       |
| i      | 8x3   | 24   |       |       |          |            |                    |          | 1       |
| j      | 8x4   | 32   |       |       |          |            | 4                  |          | 8       |
| k      | 8x6   | 48   |       |       |          |            | 1                  |          |         |
| l      | 8x8   | 64   |       |       |          |            | 1                  |          | 1       |
| m      | 8x12  | 96   |       |       |          |            |                    |          | 1       |
| n      | 9x6   | 54   |       |       |          |            |                    |          | 4       |
| o      | 10x1  | 10   |       |       |          |            |                    |          | 2       |
| p      | 10x5  | 50   |       |       |          |            |                    |          | 2       |

<small>

1. Includes both Insane and New Insane.

</small>

## Credits

### Current Maintainers

<style type="text/css">@import url(https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.css);</style>
* Subaraki, a.k.a. ArtixAllMighty
* Joel Murphy, a.k.a. Murphy's Chaos [<span class="fa fa-twitter"> </span>](http://www.twitter.com/murphyschaos) [<span class="fa fa-github"> </span>](http://www.github.com/MurphysChaos)

## Final Word

Check out [this page](https://imgur.com/gallery/R7qao). It's an article about the origins of the original Minecraft paintings.

[link_install_mod]: https://www.minecraftmods.com/how-to-install-mods-for-minecraft-forge/
[link_gibea]: https://github.com/ArtixAllMighty/Paintings--/blob/master/src/main/resources/assets/subaraki/art/gibea.png?raw=true
[link_sphax]: https://github.com/ArtixAllMighty/Paintings--/blob/master/src/main/resources/assets/subaraki/art/sphax.png?raw=true
[link_tinypics]: https://github.com/ArtixAllMighty/Paintings--/blob/master/src/main/resources/assets/subaraki/art/tinypics.png?raw=true
[link_mediumpics]: https://github.com/ArtixAllMighty/Paintings--/blob/master/src/main/resources/assets/subaraki/art/mediumpics.png?raw=true
[link_new_insane]: https://github.com/ArtixAllMighty/Paintings--/blob/master/src/main/resources/assets/subaraki/art/new_insane.png?raw=true
[link_insane]: https://github.com/ArtixAllMighty/Paintings--/blob/master/src/main/resources/assets/subaraki/art/insane.png?raw=true
[link_extended]: https://github.com/ArtixAllMighty/Paintings--/blob/master/src/main/resources/assets/subaraki/art/extended.png?raw=true
[link_massive]: https://github.com/ArtixAllMighty/Paintings--/blob/master/src/main/resources/assets/subaraki/art/massive.png?raw=true

<style type="text/css">@import url(https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.css);img{max-width:192px;}</style>


changelog.txt
Build: 1.14.4-28.0.81 - Sat Aug 31 17:57:15 GMT 2019
	cpw:
		Move indexFor and biConsumer into FMLHandshakeHandler and expose them
		publicly, so mods can add additional login messages. Closes #6087
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

	faj10:
		Added getters for FMLPlayMessages.SpawnEntity

	tobias.hotz:
		move default log level for debug file to debug instead of trace, and
		remove millis from console and GUI. Also remove unused config xmls

	cpw:
		Add in MC REGISTER nonsense. Shut up #5730
		Tweak network a bit to fix initialization order stupidity. (Constants
		don't load classes!)
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

	cpw:
		Add in some client only player connectivity events. Closes #5536
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.14.4-28.0.76 - Sat Aug 31 12:15:17 GMT 2019
	gigaherz:
		Hookup BlockEvent.FluidPlaceBlockEvent (#6103)

=========
Build: 1.14.4-28.0.75 - Fri Aug 30 03:42:51 GMT 2019
	cpw:
		Update libs to 1.0 versions, bump grgit, and fix tag search behaviour
		that resulted.
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.14.4-28.0.74 - Thu Aug 29 19:09:49 GMT 2019
	gigaherz:
		Fluid rework derp fixes. (#6098)
		* Fix nulls returning from non-null methods, and null-check of
		non-nullable values.
		* Fix missed computation of the isEmpty flag in FluidStack
		* Cleanup getAmount/isEmpty calls.
		* Remove redundant suppliers. That info is readily available from the
		vanilla Fluid / FluidState.

=========
Build: 1.14.4-28.0.73 - Thu Aug 29 04:34:22 GMT 2019
	daemon.umbra:
		Change fml.modloading.brokenfile.optifine to not imply that Forge is
		responsible for Forge-Optifine compatibility. (#6080)

	ichttt:
		Fix Digging particle sometimes using the wrong block pos to get
		additional model data (#6077)

	williewillus:
		Minor model system cleanup/documentation (#6089)

=========
Build: 1.14.4-28.0.70 - Thu Aug 29 04:28:28 GMT 2019
	ichttt:
		Add a config comment that lists the available constants of enums (#6079)

=========
Build: 1.14.4-28.0.69 - Thu Aug 29 04:18:57 GMT 2019
	bl4ckscor33:
		Restore PotionShiftEvent (#6042)

=========
Build: 1.14.4-28.0.68 - Thu Aug 29 04:13:38 GMT 2019
	danielgoossens:
		Added new InputEvent.MouseScrollEvent, fixes #5811 (#6037)

=========
Build: 1.14.4-28.0.67 - Thu Aug 29 03:34:14 GMT 2019
	luasmrks:
		Strip side annotations from CropsBlock.getSeedsItem (#6013)

	Bward7864:
		Add Villager Trading Events (#6016)

=========
Build: 1.14.4-28.0.65 - Thu Aug 29 03:14:27 GMT 2019
	wynprice999:
		Updated constants to 1.14 (#6012)

	46117728+valoeghese:
		Added hook to allow biomes to specify custom river variants. (#5969)

=========
Build: 1.14.4-28.0.63 - Thu Aug 29 01:42:25 GMT 2019
	williewillus:
		Add particle factory registration event (#6018)

=========
Build: 1.14.4-28.0.62 - Thu Aug 29 01:30:04 GMT 2019
	lex:
		Recipe system cleanup. (#6014)

	gigaherz:
		Fluid API rework (#5983) - Moved forge fluids into FluidAttributes
		companion object to the vanilla Fluid. By gigaherz - Redesigned the
		Fluid API to be closer to the Items API. By King Lemming Co-authored-by:
		King Lemming <kinglemming@gmail.com>

	gigaherz:
		Fix license headers so `gradlew build` is happy.

	gigaherz:
		Make FluidStacks non-nullable, and implement some TODOs.

=========
Build: 1.14.4-28.0.58 - Thu Aug 29 01:12:54 GMT 2019
	bjfccb:
		Fix regular arrows giving off particles after a world reload (MC-107941)
		(#6061)

	ichttt:
		Fix coremod not redirecting to the correct method (#6078)

=========
Build: 1.14.4-28.0.56 - Wed Aug 28 19:59:19 GMT 2019
	lex:
		Fix blocks being placed in adventure mode. Closes #6085 Add early error
		for setting invalid world directory. Closes #6084 Fix PlayerTickEvent
		not firing when dead, even if player isn't removed from world yet.
		Closes #6074 Fix PlayerDestroyItemEvent not firing for some cases.
		Closes #6067

=========
Build: 1.14.4-28.0.55 - Tue Aug 20 22:40:50 GMT 2019
	lex:
		Fix Javac generics breakage.

	lex:
		Sort server.properties while writing.

	lex:
		Add Argument serializer for ModIdArgument. EnumArgument can't as vanilla
		doesn't support extra data.

	lex:
		Prevent worlds from unloading while players are in out-of-world states
		by tracking all worlds references by every player in the server's player
		list.

=========
Build: 1.14.4-28.0.51 - Tue Aug 20 02:46:20 GMT 2019
	lex:
		Fix Configs not generating correctly on first load Closes #6069

	lex:
		Fix chunk load failure when entering End portal Closes #6068

=========
Build: 1.14.4-28.0.49 - Thu Aug 15 18:04:06 GMT 2019
	williewillus:
		Add hook to load custom mrls/rls not attached to a block or item. 
		(#6010)

=========
Build: 1.14.4-28.0.48 - Thu Aug 15 17:35:07 GMT 2019
	bl4ckscor33:
		Cover more cases for RightClickBlock event and restore onItemUseFirst
		(#6009)

=========
Build: 1.14.4-28.0.47 - Thu Aug 15 17:17:36 GMT 2019
	kirderf4:
		Make DimensionManager.SavedEntry save sky light consistently (#6019)

=========
Build: 1.14.4-28.0.46 - Wed Aug 14 23:13:25 GMT 2019
	lex:
		Fix extract inheritance not using proper classpath and update
		buildscript for FG asm changes.

=========
Build: 1.14.4-28.0.45 - Mon Aug 05 12:56:59 GMT 2019
	cpw:
		Grrr. Stupid error. Sorry lex: I fixed it.
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

	cpw:
		Bump modlauncher to 3.2.x. Add a sender modid and document IMC. Closes
		#5746
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

	cpw:
		Move fml's PlayerEvent and TickEvent to forge. Update `import
		net.minecraftforge.fml.common.gameevent.PlayerEvent;` to `import
		net.minecraftforge.event.entity.player.PlayerEvent;` and `import
		net.minecraftforge.fml.common.gameevent.TickEvent;` to `import
		net.minecraftforge.event.TickEvent;`
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

	cpw:
		Actually reload config when it changes on disk. Closes #5541 and #5959
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========

Paintings/Paintings.java
package Paintings;

import java.lang.reflect.Field;

import Paintings.config.ConfigFile;
import Paintings.sets.PaintingsGibea;
import Paintings.sets.PaintingsInsane;
import Paintings.sets.PaintingsNewInsane;
import Paintings.sets.PaintingsSphax;
import Paintings.sets.PaintingsTiny;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.common.FMLLog;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.Mod.EventHandler;
import net.minecraftforge.fml.common.SidedProxy;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;

@Mod(modid = Paintings.MODID, name = Paintings.NAME, version = Paintings.VERSION)
public class Paintings
{

	public static final String MODID = "morepaintings";
	public static final String VERSION = "1.10 1";
	public static final String NAME = "Paintings++";

	@SidedProxy(serverSide="Paintings.CommonProxy", clientSide="Paintings.ClientProxy")
	public static CommonProxy proxy;
	private static final String CLASS_LOC = "com.mcf.davidee.paintinggui.gui.PaintingButton";

	@EventHandler
	public void preInit(FMLPreInitializationEvent event)
	{
		ConfigFile.instance.loadConfig(event.getSuggestedConfigurationFile());

		proxy.registerRenderInformation();

	}

	String [] textureLoc= new String[5];
	int[] size = new int[5];
	boolean[] bools = new boolean[5];

	@EventHandler
	public void load(FMLInitializationEvent init)
	{
		boolean Insane = ConfigFile.instance.Insane;
		boolean sphax = ConfigFile.instance.Sphax;
		boolean tiny = ConfigFile.instance.TinyPics;
		boolean gib = ConfigFile.instance.Gibea;
		boolean nInsane = ConfigFile.instance.NewInsane;

		bools[0] = Insane;
		bools[1] = tiny;
		bools[2] = nInsane;
		bools[3] = gib;
		bools[4] = sphax;

		textureLoc[0] = "insane";
		textureLoc[1] = "tiny";
		textureLoc[2] = "newInsane";
		textureLoc[3] = "gib";
		textureLoc[4] = "sphax";

		size[0] = size[1] = size[2] = 512;
		size[3] = size[4] = 256;

		for(int i = 0; i < 5; i++){
			if(bools[i])
				try {
					Class altClass = Class.forName("com.mcf.davidee.paintinggui.gui.PaintingButton");
					paintingGuiTextureHelper(altClass, "TEXTURE", new ResourceLocation("subaraki:art/"+ textureLoc[i] +".png"));
					paintingGuiHelper(altClass, "KZ_WIDTH", size[i]);
					paintingGuiHelper(altClass, "KZ_HEIGHT", size[i]);
				} catch (Exception e) {
					FMLLog.getLogger().info("Davidees painting mod not installed or to old/new. Skipping");
				}
		}

		if ((!tiny) && (!Insane) && (!sphax) && (gib))
			PaintingsGibea.addPaintings();

		if ((!tiny) && (!Insane) && (sphax) && (!gib))
			PaintingsSphax.addPaintings();

		if ((tiny) && (!Insane) && (!sphax) && (!gib))
			PaintingsTiny.addPaintings();

		if ((!tiny) && (Insane) && (!sphax) && (!gib))
			PaintingsInsane.addPaintings();

		if(nInsane && !Insane && !gib && !tiny && !sphax)
			PaintingsNewInsane.addPaintings();

	}

	private void paintingGuiHelper(Class c, String field, int value)
			throws Exception
	{
		Field f = c.getField(field);
		f.setAccessible(true);
		f.set(null, Integer.valueOf(value));
	}

	private void paintingGuiTextureHelper(Class c, String field, ResourceLocation loc)
			throws Exception
	{
		Field f = c.getField(field);
		f.setAccessible(true);
		f.set(null, loc);
	}
}
